ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,858 @@
+
+-->
+
+This chapter discusses the concepts of the proposal and how it tackles the
+current issues with reprocessing and the updatability of the engine.
+
+The main idea of the proposal is to avoid the processing of the records on reprocessing (i.e. applying the business logic again). Instead, the broker should apply only the state changes as they were before on processing.
+
+In the domain of stream-processing/event-sourcing, this kind of rebuilding the state is known as **replay**. Comparing to reprocessing, the records (commands or events) are not processed again, only the state changes of the events are applied.
+
+In order to extract the processing from the state changes, the workflow instance
+records are split into commands and events. Commands represent the requests for
+change, while events represent that which has already happened.
+
+Currently, we already have commands and events, but the records we use today do
+not clearly exhibit these semantics. In particular, some records which we
+consider to be events actually lead to the execution of business logic to
+reproduce the state changes they represent, instead of capturing those state
+changes as something that has already happened.
+
+The following is an example of how the BPMN Element Records (which are currently
+all events) can be split into commands and events. You can read this as: when
+processing the `Element_Activating` command, we change the state and write the
+`Element_Activated` event that represents and contains these state changes.
+
+Commands | Events 
+---|---
+Element_Activating | Element_Activated 
+Element_Completing | Element_Completed
+Element_Terminating | Element_Terminated
+Event_Occur (?) | Event_Occurred (?)
+/ | Sequence_Flow_Taken
+
+> Note, that only the `Event_Occur` currently does not exist in Zeebe. It is at
+> this time not yet clear whether we need the `Event_Occur` command and/or the
+> `Event_Occurred` event.
+
+> Also note, that `Sequence_Flow_Taken` is an event without a command. We'll
+> need to find a way to make this fit the strict commands/events separation.
+
+## Processing and Replay
+
+The following diagram shows from left to right a log stream consisting of
+command and event records. Command records can only be processed, while event
+records can only be replayed.
+
+![event-sourcing](ZEP-wf-stream-processing/event-sourcing.png)
+
+> The black arrow indicates that particular events and commands are appended to
+> the log stream as a result of processing a previous command (i.e. it
+> represents the source record position - explained later in more detail). The
+> orange box wrapping some records indicates that these were appended to log
+> stream atomically.
+
+### Activities during Processing
+
+The **commands** are processed only on processing and drive the workflow execution. The processing of a command includes:
```

Comment:
I agree that we should clearly define our wording. In my opinion it should always be processing (stream processing, record processing, command processing and event processing). State changes can be applied, so perhaps it's okay to discuss applying the state changes of an event, but I would defer from using "commands are applied" because applying seems so straightforward, while command processing is not.

Reply:
It was just a suggestion, I think applying, processing, etc., it's subjective what is "straightforward" and what isn't, but you get the idea :smile:

Reply:
Mine as well 😄 we should just define terminology and use it consistently.

Reply:
I would like to go with "process commands", "replay events", "apply state changes".

Do we still need to change the sentence?

Reply:
I think the wording is still a little ambiguous for a first time reader - it sounds mostly like commands are only processed during reprocessing, but not that they are the only thing we process. So maybe we can just make it clearer that we don't process anything else. I know that's implied by reading the rest of the document, but it would help first time readers here I think.

Just my 2 cents

Reply:
@npepinpe I've had another go at this description (also note the additions to the `Processing and Replay` introduction. Please let me know whether this helps.

I've also applied the usage of "process commands", "replay events", "apply state changes" consistently throughout the document.